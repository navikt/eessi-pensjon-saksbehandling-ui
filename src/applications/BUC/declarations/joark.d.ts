
export interface JoarkFileVariant {
  variantformat:string;
  filnavn: string;
}

export interface JoarkDoc {
  dokumentInfoId: string;
  tittel: string;
  dokumentvarianter: Array<JoarkFileVariant>;
}
export interface JoarkPoster {
  journalpostId: string;
  tittel: string;
  tema: string;
  datoOpprettet: string;
  tilleggsopplysninger?: string;
  dokumenter: Array<JoarkDoc>
}

export interface JoarkFile {
  journalpostId: string;
  tittel: string;
  tema: string;
  datoOpprettet: Date;
  dokumentInfoId: string;
  tilleggsopplysninger?: string;
  variant: JoarkFileVariant
}

interface Content {
  base64: string;
}

export interface JoarkFileWithContent extends JoarkFile {
  content: Content;
  name: string;
  size: number;
  mimetype: string;
}
