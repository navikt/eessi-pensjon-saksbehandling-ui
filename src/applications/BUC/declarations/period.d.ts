export interface PeriodDate {
  day: string;
  month: string;
  year: string;
}

export interface Country {
  label: string;
  value: string;
}

export interface Period {
  id?: number,
  type: string;
  startDate: PeriodDate;
  endDate?: PeriodDate;
  dateType: string;
  uncertainDate: boolean;
  country: Country;
  comment: string;
  otherType?: string;
  insuranceName?: string;
  insuranceType?: string;
  insuranceId ?: string;
  workActivity ?: string;
  workName ?: string;
  workType ?: string;
  workStreet ?: string;
  workCity ?: string;
  workZipCode ?: string;
  workRegion ?: string;
  childFirstName ?: string;
  childLastName ?: string;
  childBirthDate ?: PeriodDate;
  learnInstitution ?: string;
  payingInstitution ?: string;
  attachments ?: Array[any];
  [key: string]: any;
}

export type Periods = Array<Period>

export type PeriodErrors = {[k: string]: string | undefined}

export interface PayloadPeriod {
  land: string;
  periode: null | {
    lukketPeriode?: {
      fom: string | null;
      tom: string | null;
    }
    openPeriode?:  {
      fom: string | null;
      extra: string;
    }
  };
  vedlegg ?: Array[any];
  trygdeordningnavn ?: string;
  medlemskap ?: string;
  forsikkringEllerRegistreringNr ?: string;
  annenInformasjon ?: string;
  usikkerDatoIndikator ?: string;
  jobbUnderAnsattEllerSelvstendig ?: string;
  navnFirma ?: string;
  typePeriode ?: string;
  adresseFirma ?: {
    postnummer: string | undefined;
    by: string | undefined;
    land: string;
    gate: string | undefined;
    region: string | undefined;
  };
  navnPaaInstitusjon ?: string;
  informasjonBarn ?:  {
    etternavn: string | null | undefined;
    fornavn: string | null | undefined;
    foedseldato: string | null | undefined;
    land: string;
  }
}

export interface Payload {
  ansattSelvstendigPerioder ?: Array<PayloadPeriod>;
  boPerioder ?: Array<PayloadPeriod>;
  barnepassPerioder ?: Array<PayloadPeriod>;
  frivilligPerioder ?: Array<PayloadPeriod>;
  forsvartjenestePerioder ?: Array<PayloadPeriod>;
  foedselspermisjonPerioder ?: Array<PayloadPeriod>;
  opplaeringPerioder ?: Array<PayloadPeriod>;
  arbeidsledigPerioder ?: Array<PayloadPeriod>;
  sykePerioder ?: Array<PayloadPeriod>;
  andrePerioder ?: Array<PayloadPeriod>;
}

export interface P4000Info {
  person: any,
  bank: any,
  stayAbroad: Array<Period>
}

export interface P4000PayloadInfo {
  personInfo: any,
  bankInfo: any,
  periodeInfo: Array<PayloadPeriod>
}
