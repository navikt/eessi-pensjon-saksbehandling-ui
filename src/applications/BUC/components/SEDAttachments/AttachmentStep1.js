import React, { useState } from 'react'
import PT from 'prop-types'
import _ from 'lodash'
import JoarkBrowser from 'components/JoarkBrowser/JoarkBrowser'
import { Nav } from 'eessi-pensjon-ui'

export const AttachmentStep1 = (props) => {
  const { files, onSubmit, t } = props
  const [mode, setMode] = useState('view')
  const [localFiles, setLocalFiles] = useState(files && _.isArray(files.joark) ? files.joark : [])

  const onLocalFileChange = (changedFiles) => {
    setLocalFiles(changedFiles)
  }

  const onSubmitJoarkFiles = (joarkFiles) => {
    if (_.isFunction(onSubmit)) {
      onSubmit({
        ...files,
        joark: joarkFiles
      })
    }
  }

  return (
    <div className='a-buc-c-sedattachments__step1'>
      <JoarkBrowser
        {...props}
        mode={mode}
        files={localFiles}
        onFilesChange={onLocalFileChange}
      />
      <div className='mt-4'>
        {mode === 'view' ? (
          <Nav.Hovedknapp
            disabled={_.isEmpty(localFiles)}
            id='a-buc-sedbody__upload-button-id'
            className='a-buc-sedbody__upload-button'
            onClick={() => setMode('confirm')}
          >
            {t('buc:form-addSelectedAttachments')}
          </Nav.Hovedknapp>
        ) : null}
        {mode === 'confirm' ? (
          <>
            <Nav.Hovedknapp
              disabled={_.isEmpty(localFiles)}
              id='a-buc-sedbody__submit-button-id'
              className='a-buc-sedbody__submit-button mr-2'
              onClick={() => onSubmitJoarkFiles(localFiles)}
            >
              {t('buc:form-submitSelectedAttachments')}
            </Nav.Hovedknapp>
            <Nav.Knapp
              id='a-buc-sedbody__cancel-button-id'
              className='a-buc-sedbody__cancel-button'
              onClick={() => setMode('view')}
            >
              {t('buc:form-selectAgainAttachments')}
            </Nav.Knapp>
          </>
        ) : null}
      </div>
    </div>
  )
}

AttachmentStep1.propTypes = {
  files: PT.object.isRequired,
  onSubmit: PT.func,
  t: PT.func.isRequired
}

export default AttachmentStep1
