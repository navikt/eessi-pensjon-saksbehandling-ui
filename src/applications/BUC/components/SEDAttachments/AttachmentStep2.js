import React from 'react'
import PT from 'prop-types'
import PDFEditor from 'applications/PDF/components/PDFEditor/PDFEditor'
import { Nav } from 'eessi-pensjon-ui'

export const AttachmentStep2 = (props) => {
  const { setStep, t } = props

  const onForwardButtonClick = () => {
    setStep('generate')
  }

  const onBackButtonClick = () => {
    setStep('select')
  }

  return (
    <div className='a-buc-c-sedattachments__step2'>
      <PDFEditor {...props} />
      <Nav.Hovedknapp
        id='a-buc-c-sedattachmnents__next-button-id'
        className='a-buc-c-sedattachmnents__next-button'
        onClick={onForwardButtonClick}
      >
        {t('ui:next')}
      </Nav.Hovedknapp>
      <Nav.Flatknapp
        id='a-buc-c-sedattachmnents__back-button-id'
        className='a-buc-c-sedattachmnents__back-button'
        onClick={onBackButtonClick}
      >
        {t('ui:back')}
      </Nav.Flatknapp>
    </div>
  )
}

AttachmentStep2.propTypes = {
  setStep: PT.func.isRequired,
  t: PT.func.isRequired
}

export default AttachmentStep2
