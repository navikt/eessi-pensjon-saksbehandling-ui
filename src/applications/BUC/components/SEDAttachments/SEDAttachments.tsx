import JoarkBrowser from 'components/JoarkBrowser/JoarkBrowser'
import { HorizontalSeparatorDiv, VerticalSeparatorDiv } from 'components/StyledComponents'
import { AttachedFiles } from 'declarations/buc'
import { AttachedFilesPropType } from 'declarations/buc.pt'
import { JoarkFiles } from 'declarations/joark'
import _ from 'lodash'
import PT from 'prop-types'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import Knapp, { Hovedknapp } from 'nav-frontend-knapper'
import TilsetteIcon from 'assets/icons/Tilsette'
import { Normaltekst } from 'nav-frontend-typografi'
import styled from 'styled-components'

export interface SEDAttachmentsProps {
  disableButtons?: boolean;
  files: AttachedFiles;
  initialMode ?: 'view' | 'confirm';
  open?: boolean;
  onFilesChange?: (f: JoarkFiles) => void;
  onOpen?: () => void;
  onSubmit?: (f: AttachedFiles) => void;
}

const SEDAttachmentsDiv = styled.div``
const FlexDiv = styled.div`
  display: flex;
`
const NormalText = styled(Normaltekst)`
  margin-left: 0.5rem;
`

const SEDAttachments: React.FC<SEDAttachmentsProps> = ({
  disableButtons = false, files, initialMode = 'view', open = false, onFilesChange, onOpen, onSubmit
}: SEDAttachmentsProps): JSX.Element => {
  const [mode, setMode] = useState<string>(initialMode)
  const [localFiles, setLocalFiles] = useState<JoarkFiles>(files && _.isArray(files.joark) ? files.joark as JoarkFiles : [])
  const { t } = useTranslation()
  const onEnableAttachmentsButtonClicked = (): void => {
    if (_.isFunction(onOpen)) {
      onOpen()
    }
  }

  const onLocalFileChange = (changedFiles: JoarkFiles): void => {
    setLocalFiles(changedFiles)
    if (_.isFunction(onFilesChange)) {
      onFilesChange(changedFiles)
    }
  }

  const onSubmitJoarkFiles = (joarkFiles: JoarkFiles): void => {
    if (_.isFunction(onSubmit)) {
      onSubmit({
        ...files,
        joark: joarkFiles
      })
    }
  }

  return (
    <SEDAttachmentsDiv data-testId='a-buc-c-sedattachments'>
      {!open ? (
        <Knapp
          data-testId='a-buc-c-sedattachments__enable-button-id'
          onClick={onEnableAttachmentsButtonClicked}
        >
          <FlexDiv>
            <TilsetteIcon />
            <NormalText>
              {t('ui:addAttachments')}
            </NormalText>
          </FlexDiv>
        </Knapp>
      ) : (
        <>
          <JoarkBrowser
            files={localFiles}
            mode={mode}
            onFilesChange={onLocalFileChange}
          />
          <VerticalSeparatorDiv data-size='1.5' />
          {mode === 'view' && (
            <Hovedknapp
              disabled={_.isEmpty(localFiles)}
              id='a-buc-c-sedattachments__upload-button-id'
              className='a-buc-c-sedattachments__upload-button'
              onClick={() => setMode('confirm')}
            >
              {t('buc:form-addSelectedAttachments')}
            </Hovedknapp>
          )}
          {mode === 'confirm' && (
            <>
              <Hovedknapp
                disabled={_.isEmpty(localFiles) || disableButtons}
                data-testId='a-buc-c-sedattachments__submit-button-id'
                onClick={() => onSubmitJoarkFiles(localFiles)}
              >
                <NormalText>
                  {t('buc:form-submitSelectedAttachments')}
                </NormalText>
              </Hovedknapp>
              <HorizontalSeparatorDiv />
              <Knapp
                disabled={disableButtons}
                data-testId='a-buc-c-sedattachments__cancel-button-id'
                className='a-buc-c-sedattachments__cancel-button'
                onClick={() => setMode('view')}
              >
                <NormalText>
                  {t('buc:form-selectAgainAttachments')}
                </NormalText>
              </Knapp>
            </>
          )}
        </>
      )}
    </SEDAttachmentsDiv>
  )
}

SEDAttachments.propTypes = {
  disableButtons: PT.bool,
  files: AttachedFilesPropType.isRequired,
  initialMode: PT.oneOf(['view', 'confirm']),
  open: PT.bool,
  onFilesChange: PT.func,
  onOpen: PT.func,
  onSubmit: PT.func
}

export default SEDAttachments
