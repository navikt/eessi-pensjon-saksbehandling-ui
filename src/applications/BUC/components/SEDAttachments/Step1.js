import React from 'react'
import PT from 'prop-types'
import _ from 'lodash'
import classNames from 'classnames'
import JoarkBrowser from 'components/ui/JoarkBrowser/JoarkBrowser'
import FileUpload from 'components/ui/FileUpload/FileUpload'
import { Ingress, Hovedknapp } from 'components/ui/Nav'

const Step1 = (props) => {
  const { t, files, setFiles, setStep } = props

  const onEditButtonClick = () => {
    setStep('edit')
  }

  const handleFileChange = (changedFiles, key) => {
    let newFiles = _.cloneDeep(files)
    newFiles[key] = changedFiles
    setFiles(newFiles)
  }

  const getFiles = (key) => {
    if (files.hasOwnProperty(key)) {
      return (files[key])
    }
    return []
  }

  const handleJoarkFilesChange = (joarkFiles) => {
    handleFileChange(joarkFiles, 'joark')
  }

  const handleFileUploadFileChange = (fileUploadFiles) => {
    handleFileChange(fileUploadFiles, 'fileupload')
  }

  const handleBeforeDrop = () => {
  }

  const handleAfterDrop = () => {
  }

  return <div>
    <Ingress className='mt-4 mb-3'>{t('buc:form-addAttachmentsFromJOARK')}</Ingress>
    <JoarkBrowser
      {...props}
      files={getFiles('joark')}
      onFilesChange={handleJoarkFilesChange} />
    <Ingress className='mt-4 mb-3'>{t('buc:form-addAttachmentsFromFileUpload')}</Ingress>
    <FileUpload t={t}
      fileUploadDroppableId={'SEDAttachments'}
      id={classNames('a-buc-c-sedattachments-fileupload-id', 'mb-3')}
      className={classNames('fileUpload', 'mb-3')}
      accept={['application/pdf', 'image/jpeg', 'image/png']}
      files={files.hasOwnProperty('fileupload') ? files['fileupload'] : []}
      beforeDrop={handleBeforeDrop}
      afterDrop={handleAfterDrop}
      onFileChange={handleFileUploadFileChange} />
    <Hovedknapp
      id='a-buc-c-sedattachmnents__edit-button-id'
      className='a-buc-c-sedattachmnents__edit-button'
      onClick={onEditButtonClick}>
      {t('ui:edit')}
    </Hovedknapp>
  </div>
}

Step1.propTypes = {
  actions: PT.object,
  t: PT.func.isRequired,
  setFiles: PT.func.isRequired
}

export default Step1
