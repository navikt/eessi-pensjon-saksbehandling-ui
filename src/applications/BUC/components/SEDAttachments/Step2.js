import React, { useState, useEffect } from 'react'
import PT from 'prop-types'
import _ from 'lodash'
import classNames from 'classnames'
import PDFEditor from 'applications/PDF/components/PDFEditor/PDFEditor'
import { Ingress, Hovedknapp, Flatknapp } from 'components/ui/Nav'

const Step2 = (props) => {
  const { t, files, setFiles, setStep } = props

  const onForwardButtonClick = () => {
    setStep('generate')
  }

  const onBackButtonClick = () => {
    setStep('select')
  }

  return <div>
    <PDFEditor {...props} />
    <Hovedknapp
      id='a-buc-c-sedattachmnents__next-button-id'
      className='a-buc-c-sedattachmnents__next-button'
      onClick={onForwardButtonClick}>
      {t('ui:next')}
    </Hovedknapp>
    <Flatknapp
      id='a-buc-c-sedattachmnents__next-button-id'
      className='a-buc-c-sedattachmnents__next-button'
      onClick={onBackButtonClick}>
      {t('ui:back')}
    </Flatknapp>
  </div>
}

Step2.propTypes = {
  actions: PT.object,
  t: PT.func.isRequired,
  setFiles: PT.func.isRequired
}

export default Step2
