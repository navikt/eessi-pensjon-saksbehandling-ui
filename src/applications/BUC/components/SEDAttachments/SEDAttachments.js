import React, { useState, useEffect } from 'react'
import PT from 'prop-types'
import _ from 'lodash'
import classNames from 'classnames'
import { Stegindikator, Checkbox, Hovedknapp } from 'components/ui/Nav'
import Step1 from './Step1'
import Step2 from './Step2'

const SEDAttachments = (props) => {
  const { t, files, setFiles } = props
  const [ step, setStep ] = useState(0)
  const [ enableAttachments, setEnableAttachments ] = useState(false)

  const onChange = () => {
    setEnableAttachments(!enableAttachments)
  }

  return <div className='a-buc-c-sedattachmnents'>
    <Checkbox label={t('buc:form-enableAttachments')}
      onChange={onChange}
      checked={enableAttachments} />
    {enableAttachments && step === 1
      ? <Stegindikator
        visLabel
        autoResponsiv
        steg={[
          { label: t('buc:attachments-step0'), aktiv: (step === 0) },
          { label: t('buc:attachments-step1'), aktiv: (step === 1) },
          { label: t('buc:attachments-step2'), aktiv: (step === 2) }
        ]} /> : null}
    {enableAttachments && step === 0 ? <Step1 setStep={setStep} {...props} /> : null}
    {enableAttachments && step === 1 ? <Step2 setStep={setStep} {...props} /> : null}
  </div>
}

SEDAttachments.propTypes = {
  actions: PT.object,
  t: PT.func.isRequired,
  setFiles: PT.func.isRequired
}

export default SEDAttachments
