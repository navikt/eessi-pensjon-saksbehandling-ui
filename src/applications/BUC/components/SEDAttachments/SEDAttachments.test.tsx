import { JoarkFile } from 'declarations/joark'
import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import mockJoarkReduced from 'mocks/joark/joarkReduced'
import SEDAttachments, { SEDAttachmentsProps } from './SEDAttachments'

jest.mock('components/JoarkBrowser/JoarkBrowser', () =>
  (props: any) => (<div className='mock-joarkbrowser' onClick={() => props.onFilesChange([{ foo: 'bar' }])} />)
)

describe('applications/BUC/components/SEDAttachments/SEDAttachments', () => {
  let wrapper: ReactWrapper
  const initialMockProps: SEDAttachmentsProps = {
    files: {},
    onFilesChange: jest.fn(),
    open: true,
    onOpen: jest.fn(),
    onSubmit: jest.fn()
  }

  beforeEach(() => {
    wrapper = mount(<SEDAttachments {...initialMockProps} />)
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('Renders', () => {
    expect(wrapper.isEmptyRender()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('Has proper HTML structure', () => {
    wrapper = mount(<SEDAttachments {...initialMockProps} open={false} />)
    expect(wrapper.exists('.a-buc-c-sedattachments')).toBeTruthy()
    expect(wrapper.exists('#a-buc-c-sedattachments__enable-button-id')).toBeTruthy()
  })

  it('onFilesChanged triggered', () => {
    (initialMockProps.onFilesChange as jest.Mock).mockReset()
    wrapper.find('.mock-joarkbrowser').simulate('click')
    expect(initialMockProps.onFilesChange).toHaveBeenCalledWith([{ foo: 'bar' }])
  })

  it('onSubmit triggered', () => {
    (initialMockProps.onOpen as jest.Mock).mockReset()
    wrapper = mount(<SEDAttachments {...initialMockProps} files={{ joark: [(mockJoarkReduced[0] as JoarkFile)] }} />)
    // @ts-ignore
    wrapper.find('button.a-buc-c-sedattachments__submit-button').props().onClick()
    expect(initialMockProps.onSubmit).toHaveBeenCalledWith({ joark: [(mockJoarkReduced[0] as JoarkFile)] })
  })

  it('onEnableAttachmentsButtonClicked triggered', () => {
    (initialMockProps.onOpen as jest.Mock).mockReset()
    wrapper = mount(<SEDAttachments {...initialMockProps} open={false} />)
    // @ts-ignore
    wrapper.find('Knapp.a-buc-c-sedattachments__enable-button').props().onClick()
    expect(initialMockProps.onOpen).toHaveBeenCalled()
  })
})
