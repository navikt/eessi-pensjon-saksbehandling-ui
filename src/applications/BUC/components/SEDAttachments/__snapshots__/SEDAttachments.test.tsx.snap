// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applications/BUC/components/SEDAttachments/SEDAttachments Renders 1`] = `
<SEDAttachments
  files={Object {}}
  initialMode="view"
  onFilesChange={[MockFunction]}
  onOpen={[MockFunction]}
  onSubmit={[MockFunction]}
  open={true}
>
  <styled.div
    data-testId="a-buc-c-sedattachments"
  >
    <div
      className="sc-AxiKw bOTXee"
      data-testId="a-buc-c-sedattachments"
    >
      <Component
        files={Array []}
        mode="view"
        onFilesChange={[Function]}
      >
        <div
          className="mock-joarkbrowser"
          onClick={[Function]}
        />
      </Component>
      <styled.div
        data-size="1.5"
      >
        <div
          className="sc-AxjAm fIaKoH"
          data-size="1.5"
        />
      </styled.div>
      <Hovedknapp
        className="a-buc-c-sedattachments__upload-button"
        disabled={true}
        id="a-buc-c-sedattachments__upload-button-id"
        onClick={[Function]}
      >
        <KnappBase
          autoDisableVedSpinner={false}
          className="a-buc-c-sedattachments__upload-button"
          disabled={true}
          htmlType="submit"
          id="a-buc-c-sedattachments__upload-button-id"
          inaktivKlasseVedDisabled={false}
          kompakt={false}
          mini={false}
          onClick={[Function]}
          spinner={false}
          type="hoved"
        >
          <button
            className="knapp a-buc-c-sedattachments__upload-button knapp--hoved knapp--disabled"
            disabled={true}
            id="a-buc-c-sedattachments__upload-button-id"
            onClick={[Function]}
            type="submit"
          >
            buc:form-addSelectedAttachments
          </button>
        </KnappBase>
      </Hovedknapp>
    </div>
  </styled.div>
</SEDAttachments>
`;
